### ISSUE TITLE
OpenMM and Amber handle virtual sites very differently (for TIP[45]P)

### ISSUE THREAD

**Issue by swails** [pandegroup/openmm#794]
In particular, Amber transfers the entire force on the EP to the parent atom and transfers the torques throughout the whole frame.  By contrast, OpenMM treats TIP4P EPs as a `ThreeParticleAverageSite`, so the EP forces are simply multiplied by their weights when transferred to each particle, which *seems* to be what GROMACS does according to [this article](http://pubs.acs.org/doi/full/10.1021/ct900549r).

This manifests itself as very different forces coming out of both Amber and OpenMM.  For example, in a pure TIP4P water box.  I put the script that demonstrates [this in a gist](https://gist.github.com/swails/f1baff0db7ec1c3cbd25).

The prmtop and inpcrd files are in the ParmEd github repository [the prmtop file](https://github.com/swails/ParmEd/blob/master/test/files/tip4p.parm7) and [the restart file](https://github.com/swails/ParmEd/blob/master/test/files/tip4p.rst7).

The Amber forces were computed using a new sander Python API which won't be released until AmberTools 15 in April(ish).  So I put the sander forces [in another gist](https://gist.github.com/swails/225b559eead7ff0b2dac).  Of course the root cause of the difference is pretty obvious, so I didn't think anybody actually had to _run_ this example...

Just to give you an idea of what I'm seeing (all forces are kcal/mol/A):

For the tip4p system (`forces2` comes from OpenMM, `f` comes from sander):

```Python
In [68]: forces2[:3]
Out[68]: 
array([[  8.26955124,  26.89731116, -17.19372798],
       [  7.32330273,  -0.94056631,   8.49455393],
       [-19.87511825, -18.05324864,   5.71479643]])

In [69]: f[:3]
Out[69]: 
array([[  9.14448733,  33.97355293, -23.85368691],
       [  6.80433206,  -4.79563369,  12.03567012],
       [-20.21538842, -21.24630844,   8.75403629]])

In [70]: forces2[:3].sum(axis=0)
Out[70]: array([-4.28226427,  7.90349621, -2.98437762])

In [71]: f[:3].sum(axis=0)
Out[71]: array([-4.26656903,  7.93161081, -3.06398051])
```

By contrast, the agreement between sander and OpenMM with systems without EPs is quite good:

```Python
In [86]: forces2[:3]
Out[86]: 
array([[  0.20008624,  36.41765425, -17.91502895],
       [  2.06049683,  -1.71298121,  -2.86791275],
       [  3.21448733,   7.27584675,   1.73497781]])

In [87]: f[:3]
Out[87]: 
array([[  0.21078733,  36.41600782, -17.90203275],
       [  2.05975383,  -1.70862685,  -2.87636709],
       [  3.22186205,   7.27746535,   1.74203294]])
...
In [89]: forces2[-3:]
Out[89]: 
array([[  5.98226441,   4.12942912,  13.56357698],
       [ -0.62544022,  -0.1941101 ,  -7.9997338 ],
       [ -9.19320889,  -5.54310523, -10.14606775]])

In [90]: f[-3:]
Out[90]: 
array([[  5.97395607,   4.13476559,  13.56377223],
       [ -0.63227084,  -0.19987419,  -7.99790922],
       [ -9.17905762,  -5.54921716, -10.13269815]])
```

Does any of this matter?  And if so, which is 'correct' (and why)?

**Comment by peastman** [comment]
I'm pretty sure if you really give them identical systems they'll produce identical results.  There's a very clear correct answer: F = -dE/dx.  If you don't satisfy that, you won't conserve energy.  There are lots of ways to define a virtual site, and the calculations for redistributing the forces are very different.  But if two definitions are really equivalent (they always produce identical virtual site positions), then the forces will necessarily come out the same.

**Comment by swails** [comment]
So the `sum` reduction I did showed that the _net_ force on each water molecule is the same in each dimension.  So I would naively expect both codes to conserve total momentum (a good thing).  It's not clear to me that both conserve angular momentum, though (I'm not sure which one, if either, does).  Although, IIRC, even PME doesn't strictly conserve angular momentum and we seem to be OK with that.  I'll try to do some more systematic studies of OpenMM vs. Amber with TIP4PEw and see if there are any measurable differences (others are welcome to as well).

> There's a very clear correct answer: F = -dE/dx. If you don't satisfy that, you won't conserve energy.

Except energy conservation is only one conservation law that we care about.  And I think Newton's law is a bit ambiguous here, since you can't force on a massless particle.

> But if two definitions are really equivalent (they always produce identical virtual site positions), then the forces will necessarily come out the same.

That's what I figured.  But both Amber and OpenMM keep the TIP4P virtual site in the same place, but the forces on each (real) atom are very clearly different.  But the _net_ force in each dimension on the frame as a whole does match between the two codes.  My knowledge of virtual sites is admittedly too novice to know whether or not I should be concerned about that.

**Comment by jlmaccal** [comment]
A couple of thoughts:
1. No simulation with PBC conserves total angular momentum, PME or not.
2. The v-site constructions should conserve the torque, which will be applied to the atoms used to construct the v-site

> Amber... transfers the torques throughout the whole frame

What do you mean by this?

**Comment by swails** [comment]
> Amber... transfers the torques throughout the whole frame
> What do you mean by this?

It projects the torques onto each of the atoms in the frame using the chain rule.  I *don't* think that OpenMM conserves torque in `ThreePointAverageSite`.

Do you have a reference for how VSites are supposed to work?  The GROMACS one I saw seems to do things the OMM way, which doesn't seem to conserve torque.  The one referenced in the TIP4PEw paper (out of Teresa Head-Gordon's lab) in turn links to another paper from Berendsen and van Gunsteren (among others) from 1984 that I simply can't get my hands on...

**Comment by jlmaccal** [comment]
> another paper from Berendsen and van Gunsteren (among others) from 1984 that I simply can't get my hands on...

Yes, this is a book chapter that seems quite hard to find.

> It projects the torques onto each of the atoms in the frame using the chain rule.

This sounds similar to how v-sites work. The position of the v-site is computed as a function of other "real" atoms. The forces on the v-sites are then transferred to the real atoms using the chain rule. In the case of the `ThreePointAverageSite`, this is just a weighted partitioning of the force between the three real atoms.

**Comment by jlmaccal** [comment]
> which doesn't seem to conserve torque

What makes you say that?

If we have 3 particles, r1, r2, r3, and a v-site, where rv = w1 r1 + w2 r2 + w3 r3, then the total torque is:

T = r1 x F1 + r2 x F2 + r3 x F3 + (w1 r1 + w2 r2 + w3 r3) x Fv

Because the cross product is distributive, we can write this as:

T = r1 x F1 + w1 r1 x Fv + r2 x F2 + w2 r2 x Fv + r3 x F3 + w3 r3 x Fv

Or:

T = r1 x (F1 + w1 Fv) + r2 x (F2 + w2 Fv) + r3 x (F3 + w3 Fv).

This last expression corresponds to how the forces get redistributed by `ThreePointAverageSite`, which shows that the total torque is conserved.

**Comment by peastman** [comment]
As long as you're calculating the forces correctly (as the negative gradient of the energy), the simulation will preserve all conservation laws of the underlying system.  A periodic box doesn't have a conserved angular momentum because the system isn't invariant under rotations.  But an isolated water molecule is invariant under rotations, so it does have a conserved angular momentum.

In any case, there is an unambiguous "correct" value for the force on every atom: the negative gradient of the energy.  However you do the calculation, it needs to produce that result in the end.

**Comment by swails** [comment]
OK, I'm satisfied.  I should have followed the torque derivation all the way through -- it was pretty straightforward to demonstrate conservation (I figured, incorrectly, that since Amber's expression _did_ conserve torque, that OpenMM's must not since the resulting forces were different... a poor assumption to be sure).  As TIP4P is rigid and the total force on the whole molecule is consistent for both codes, I don't suspect I'd be able to find any differences in computed quantities, so I'll push this to a lower priority.

Closing this for now.  Thanks @peastman and @jlmaccal for the discussion!
